File: coolero/services/liquidctl_device_extractors/commander_pro.py
#  Coolero - monitor and control your cooling and other devices
#  Copyright (c) 2021  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging
from typing import List, Dict, Any, Tuple

from liquidctl.driver.commander_pro import CommanderPro

from coolero.models.channel_info import ChannelInfo
from coolero.models.device_info import DeviceInfo
from coolero.models.lighting_mode import LightingMode
from coolero.models.speed_options import SpeedOptions
from coolero.models.status import TempStatus, ChannelStatus
from coolero.services.liquidctl_device_extractors import LiquidctlDeviceInfoExtractor

_LOG = logging.getLogger(__name__)


# pylint: disable=protected-access
class CommanderProExtractor(LiquidctlDeviceInfoExtractor):
    supported_driver = CommanderPro
    _channels: Dict[str, ChannelInfo] = {}
    _lighting_speeds: List[str] = []
    _min_temp = 20
    _max_temp = 60  # Check driver for more info. Basically >60 is considered unsafe, so we use the safe default max
    _lighting_modes: List[Tuple[str, int, int, bool, bool]] = [
        # Mode Info:
        # name, min_colors, max_colors, speed_enabled, direction_enabled
        ('off', 0, 0, False, False),
        ('fixed', 1, 1, False, False),
        ('color_shift', 0, 2, True, True),
        ('color_pulse', 0, 2, True, True),
        ('color_wave', 0, 2, True, True),
        ('visor', 0, 2, True, True),
        ('blink', 0, 2, True, True),
        ('marquee', 0, 1, True, True),
        ('sequential', 0, 1, True, True),
        ('rainbow', 0, 0, True, True),
        ('rainbow2', 0, 0, True, True)
    ]

    @classmethod
    def extract_info(cls, device_instance: CommanderPro) -> DeviceInfo:
        for channel_name in device_instance._fan_names:
            cls._channels[channel_name] = ChannelInfo(
                speed_options=SpeedOptions(
                    min_duty=0,
                    max_duty=100,
                    profiles_enabled=True,
                    fixed_enabled=True
                )
            )

        for channel_name in device_instance._led_names:
            cls._channels[channel_name] = ChannelInfo(
                lighting_modes=cls._get_filtered_color_channel_modes(channel_name)
            )

        cls._lighting_speeds = ['slow', 'medium', 'fast']

        return DeviceInfo(
            channels=cls._channels,
            lighting_speeds=cls._lighting_speeds,
            temp_min=cls._min_temp,
            temp_max=cls._max_temp,
            profile_max_length=6,
            profile_min_length=2
        )

    @classmethod
    def _get_filtered_color_channel_modes(cls, channel_name: str) -> List[LightingMode]:
        channel_modes = []
        for mode_name, min_colors, max_colors, speed_enabled, direction_enabled in cls._lighting_modes:
            channel_modes.append(LightingMode(
                mode_name, cls._channel_to_frontend_name(mode_name),
                min_colors, max_colors,
                speed_enabled, direction_enabled
            ))
        return channel_modes

    @classmethod
    def _get_temperatures(cls, status_dict: Dict[str, Any], device_id: int) -> List[TempStatus]:
        probes = cls._get_temp_probes(status_dict)
        return [
            TempStatus(name, temp, name.capitalize(), f'LC#{device_id} {name.capitalize()}')
            for name, temp in probes
        ]

    @classmethod
    def _get_channel_statuses(cls, status_dict: Dict[str, Any]) -> List[ChannelStatus]:
        channel_statuses: List[ChannelStatus] = []
        multiple_fans_rpm = cls._get_multiple_fans_rpm(status_dict)
        channel_statuses.extend(
            ChannelStatus(name, rpm=rpm)
            for name, rpm in multiple_fans_rpm
        )
        return channel_statuses

File: coolercontrold/src/repositories/liquidctl/supported_devices/commander_pro.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2021-2025  Guy Boldon, Eren Simsek and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::collections::HashMap;

use crate::device::{ChannelInfo, DeviceInfo, DriverInfo, DriverType, LightingMode, SpeedOptions};
use crate::repositories::liquidctl::base_driver::BaseDriver;
use crate::repositories::liquidctl::liqctld_client::DeviceResponse;
use crate::repositories::liquidctl::supported_devices::device_support::{ColorMode, DeviceSupport};

#[derive(Debug)]
pub struct CommanderProSupport;
// commander_pro.py

impl CommanderProSupport {
    pub fn new() -> Self {
        Self {}
    }
}

impl DeviceSupport for CommanderProSupport {
    fn supported_driver(&self) -> BaseDriver {
        BaseDriver::CommanderPro
    }

    fn extract_info(&self, device_response: &DeviceResponse) -> DeviceInfo {
        let mut channels = HashMap::new();
        for channel_name in &device_response.properties.speed_channels {
            channels.insert(
                channel_name.to_owned(),
                ChannelInfo {
                    speed_options: Some(SpeedOptions {
                        min_duty: 0,
                        max_duty: 100,
                        // Internal profiles for the commander pro only work with RPMs! not duty %
                        profiles_enabled: false,
                        fixed_enabled: true,
                        manual_profiles_enabled: true,
                    }),
                    ..Default::default()
                },
            );
        }
        for channel_name in &device_response.properties.color_channels {
            let lighting_modes = self.get_color_channel_modes(None);
            channels.insert(
                channel_name.to_owned(),
                ChannelInfo {
                    lighting_modes,
                    ..Default::default()
                },
            );
        }
        let lighting_speeds = vec!["slow".to_string(), "medium".to_string(), "fast".to_string()];
        DeviceInfo {
            channels,
            lighting_speeds,
            temp_min: 20,
            temp_max: 60,
            driver_info: DriverInfo {
                drv_type: DriverType::Liquidctl,
                name: Some(self.supported_driver().to_string()),
                version: device_response.liquidctl_version.clone(),
                locations: self.collect_driver_locations(device_response),
            },
            ..Default::default()
        }
    }

    fn get_color_channel_modes(&self, _channel_name: Option<&str>) -> Vec<LightingMode> {
        let color_modes = vec![
            ColorMode::new("off", 0, 0, false, false),
            ColorMode::new("fixed", 1, 1, false, false),
            ColorMode::new("color_shift", 0, 2, true, true),
            ColorMode::new("color_pulse", 0, 2, true, true),
            ColorMode::new("color_wave", 0, 2, true, true),
            ColorMode::new("visor", 0, 2, true, true),
            ColorMode::new("blink", 0, 2, true, true),
            ColorMode::new("marquee", 0, 1, true, true),
            ColorMode::new("sequential", 0, 1, true, true),
            ColorMode::new("rainbow", 0, 0, true, true),
            ColorMode::new("rainbow2", 0, 0, true, true),
        ];
        self.convert_to_channel_lighting_modes(color_modes)
    }
}
